### PR TITLE
Add codeowners feature

### LINKED ISSUE #10161: support Github-like CODEOWNERS?
We are a team of ~15 engineers using gitea for daily development.

What is the attitude of this project towards supporting the github-like CODEOWNERS system or something like it?

To be honest, I'd find this functionality very useful.

### PR DESCRIPTION
**Pull request by cl-bvl**
Hello.
This PR adds a github like configuration for the CODEOWNERS file.

Resolves: #10161

### CODE REVIEW

File: modules/git/repo_show.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (techknowlogick) [inline modules/git/repo_show.go]:
```suggestion
// Copyright 2023 The Gitea Authors. All rights reserved.
```

File: services/pull/pull.go [hunk 2]
```diff
@@ -838,3 +866,158 @@ func IsHeadEqualWithBranch(ctx context.Context, pr *issues_model.PullRequest, br
...
 	return baseCommit.HasPreviousCommit(headCommit.ID)
 }
+
+// GetCodeOwners returns the code owners configuration
+// Return empty slice if files missing
+// Return error on file system errors
+// We're trying to do the best we can when parsing a file.
+// Invalid lines are skipped. Non-existent users and groups too.
+func GetCodeOwners(ctx context.Context, repo *repo_model.Repository, branch string) ([]*CodeOwnerRule, error) {
+	files := []string{"CODEOWNERS", "docs/CODEOWNERS", ".gitea/CODEOWNERS"}
```
Comment (techknowlogick) [inline services/pull/pull.go]: Thanks for this PR <3 Love the comments, and that docs are included too.

Could you add `.github/CODEOWNERS` here too to ensure compatibility when migrating from GH to Gitea.

File: services/pull/pull.go [hunk 3]
```diff
@@ -838,3 +866,158 @@ func IsHeadEqualWithBranch(ctx context.Context, pr *issues_model.PullRequest, br
...
+
+	rules := make([]*CodeOwnerRule, 0)
+	lines := strings.Split(string(data), "\n")
+
+	for _, line := range lines {
+		tokens := tokenizeCodeOwnersLine(line)
+		if len(tokens) == 0 {
+			continue
+		} else if len(tokens) < 2 {
+			log.Info("Incorrect codeowner line: %s", line)
```
Comment (a1012112796) [inline services/pull/pull.go]:
```suggestion
```

I don't think log at here is necessary

File: services/pull/pull.go [hunk 4]
```diff
@@ -838,3 +866,158 @@ func IsHeadEqualWithBranch(ctx context.Context, pr *issues_model.PullRequest, br
...
+	for _, char := range line {
+		if escape {
+			token += string(char)
+			escape = false
+		} else if string(char) == "\\" {
+			escape = true
+		} else if string(char) == "#" {
+			break
+			//} else if string(char) == "/" {
+			//	token += "\\/"
```
Comment (a1012112796) [inline services/pull/pull.go]: ?

File: modules/git/repo_show.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+func (repo *Repository) GetFileContent(rev, path string) ([]byte, error) {
+	cmd := NewCommand(repo.Ctx, "show")
+	cmd.AddDynamicArguments(fmt.Sprintf("%s:%s", rev, path))
+	stdout, _, err := cmd.RunStdBytes(&RunOpts{Dir: repo.Path})
+	if err != nil {
+		return nil, err
+	}
+
+	return stdout, nil
+}
```
Comment (a1012112796) [inline modules/git/repo_show.go]: why not use already exist function to read repo file? (like read file logic in renderFile)

File: services/pull/pull.go [hunk 6]
```diff
@@ -123,6 +124,33 @@ func NewPullRequest(ctx context.Context, repo *repo_model.Repository, pull *issu
 		}
 
 		_, _ = issue_service.CreateComment(ctx, ops)
+
+		if coRules, err := GetCodeOwners(ctx, repo, pr.BaseBranch); err == nil {
```
Comment (a1012112796) [inline services/pull/pull.go]:
```suggestion
		if coRules, err := GetCodeOwners(ctx, repo, pr.BaseBranch); err == nil && len(coRules) > 0 {
```

File: services/pull/pull.go (hunk 2 above)
Comment (cl-bvl) [reply services/pull/pull.go]: I added it in first revision, but removed it.
Our file has a slightly different regular expression format and in any case there will be other users / groups.
Therefore, I think that maintaining the github file is unnecessary and can create additional confusion.

As example Gitlab does not support `.github` file too.
https://docs.gitlab.com/ee/user/project/codeowners/

File: services/pull/pull.go (hunk 3 above)
Comment (cl-bvl) [reply services/pull/pull.go]: Without logs, it's hard to debug why the line is being silently ignored.

File: modules/git/repo_show.go (hunk 5 above)
Comment (cl-bvl) [reply modules/git/repo_show.go]: I looking for existing method to view files, but does not find it.
Can you please help me to find it.

File: services/pull/pull.go (hunk 6 above)
Comment (cl-bvl) [reply services/pull/pull.go]: Added

File: services/pull/pull.go (hunk 4 above)
Comment (cl-bvl) [reply services/pull/pull.go]: Removed

File: services/pull/pull.go (hunk 2 above)
Comment (techknowlogick) [reply services/pull/pull.go]: That's reasonable. Thanks for clarification :)

File: docs/content/doc/usage/code-owners.en-us.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+Gitea maintains code owner files. It looks for it in the following locations in this order:
+
+- `./CODEOWNERS`
+- `./docs/CODEOWNERS`
+- `./.gitea/CODEOWNERS`
+
+And stops at the first found file.
+
+Example file:
```
Comment (lunny) [inline docs/content/doc/usage/code-owners.en-us.md:33]: What's the syntax description?

File: services/pull/pull.go (hunk 3 above)
Comment (KN4CK3R) [reply services/pull/pull.go]: `log.Debug` then?

File: modules/git/repo_show.go (hunk 5 above)
Comment (KN4CK3R) [reply modules/git/repo_show.go]:
```
commit, err := gitRepo.GetBranchCommit(repo.DefaultBranch)
blob, err := commit.GetBlobByPath("path/to/file")
bytes, err := blob.GetBlobContent() // or blob.DataAsync() to use streaming
```

File: services/pull/pull.go (hunk 3 above)
Comment (a1012112796) [reply services/pull/pull.go]: looks you want user know whether the `codeowner` file has wrong format, isn't it? but this log is a report for server manager, not user. I think you can add a format check in file view like this : https://github.com/go-gitea/gitea/pull/24267

File: modules/git/repo_show.go (hunk 5 above)
Comment (cl-bvl) [reply modules/git/repo_show.go]: Thank you.

File: docs/content/doc/usage/code-owners.en-us.md (hunk 7 above)
Comment (cl-bvl) [reply docs/content/doc/usage/code-owners.en-us.md:33]: Added

File: services/pull/pull.go (hunk 3 above)
Comment (cl-bvl) [reply services/pull/pull.go]: Thanks for good suggestion. Added

File: services/pull/pull.go [hunk 8]
```diff
@@ -838,3 +865,175 @@ func IsHeadEqualWithBranch(ctx context.Context, pr *issues_model.PullRequest, br
...
+	return rules, warnings
+}
+
+type CodeOwnerRule struct {
+	Rule     *regexp.Regexp
+	Negative bool
+	Users    []*user_model.User
+}
+
+func parseCodeOwnersLine(ctx context.Context, tokens []string) (*CodeOwnerRule, []string) {
```
Comment (wxiaoguang) [inline services/pull/pull.go]: IMO it needs some tests to cover the code, to make sure the logic won't be broken by future changes.

File: routers/web/repo/view.go [hunk 9]
```diff
@@ -361,6 +362,14 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 		if workFlowErr != nil {
 			ctx.Data["FileError"] = ctx.Locale.Tr("actions.runs.invalid_workflow_helper", workFlowErr.Error())
 		}
+	} else if util.SliceContains([]string{"CODEOWNERS", "docs/CODEOWNERS", ".gitea/CODEOWNERS"}, ctx.Repo.TreePath) {
```
Comment (lunny) [inline routers/web/repo/view.go:365]: Why it's `util.SliceContains` but not equal?

File: services/pull/check.go [hunk 10]
```diff
@@ -357,14 +357,15 @@ func testPR(id int64) {
...
-	}
+	//if err := TestPatch(pr); err != nil {
+	//	log.Error("testPatch[%-v]: %v", pr, err)
+	//	pr.Status = issues_model.PullRequestStatusError
+	//	if err := pr.UpdateCols("status"); err != nil {
+	//		log.Error("update pr [%-v] status to PullRequestStatusError failed: %v", pr, err)
+	//	}
+	//	return
+	//}
+	TestPatch(pr)
```
Comment (lunny) [inline services/pull/check.go]: Why comment the code?

File: services/pull/pull.go [hunk 11]
```diff
@@ -838,3 +865,175 @@ func IsHeadEqualWithBranch(ctx context.Context, pr *issues_model.PullRequest, br
...
+	for _, file := range files {
+		if blob, err := commit.GetBlobByPath(file); err == nil {
+			data, err = blob.GetBlobContent()
+			if err == nil {
+				break
+			}
+		}
+	}
+
+	rules, warnings := GetCodeOwnersFromContent(ctx, data)
```
Comment (lunny) [inline services/pull/pull.go]: You need to hanlde `if len(data) == 0`

**Comment by lunny** [comment]
- [ ] We need to check who can create the CodeOwner file. As GH's docs, admin permission is required.
- [ ] Reviewers could also be team
- [ ] When to send review requests to reviewers. A draft doesn't need to be reviewed immediately. Only a normal pull request or a pull request converted from a draft need to notify reviewers.

File: services/pull/pull.go (hunk 11 above)
Comment (cl-bvl) [reply services/pull/pull.go]: No, we checking for `len(data)` in `GetCodeOwnersFromContent()` func.

File: services/pull/check.go (hunk 10 above)
Comment (cl-bvl) [reply services/pull/check.go]: Mistake, reverted it.

File: routers/web/repo/view.go (hunk 9 above)
Comment (wxiaoguang) [reply routers/web/repo/view.go:365]: It means `ctx.Repo.TreePath=="CODEOWNERS" || ctx.Repo.TreePath==".gitea/CODEOWNERS" || ....`

**Comment by cl-bvl** [comment]
@lunny 
> We need to check who can create the CodeOwner file. As GH's docs, admin permission is required.
To create CODEOWNERS write permission to default branch is required. I think it's enough.
>Reviewers could also be team
Done.
>  When to send review requests to reviewers. A draft doesn't need to be reviewed immediately. Only a normal pull request or a pull request converted from a draft need to notify reviewers.
Done.

File: services/pull/pull.go (hunk 8 above)
Comment (cl-bvl) [reply services/pull/pull.go]: Added test for file parser

**Review by wxiaoguang — APPROVED** [review]
Overall LGTM. Some nits (eg: import order, `GetBlobContent` for large file would DoS the server) could be easily fixed.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
lint error is related

**Comment by lunny** [comment]
Sent 7ddd65474e6bb252e87af21e08bb735d0eef49bc to fix th lint

**Comment by silverwind** [comment]
@cl-bvl can you allow push access to your branch?

**Comment by wxiaoguang** [comment]
I guess that's GitHub's limitation, GitHub doesn't work that way. Since this PR is from a org (WinnerSoftLab), so no "Allow maintainer edit". Only the author can update the PR unless someone is added as the fork's collaborator.

**Comment by silverwind** [comment]
So effectively, this can only be merged manually on the CLI then.

In any case, lint needs to be fixed first.

**Comment by denyskon** [comment]
> So effectively, this can likely only be merged manually on the CLI.
> 
> In any case, lint needs to be fixed first.

What about https://github.com/go-gitea/gitea/pull/24910#issuecomment-1575420444 ?

**Comment by lunny** [comment]
> I guess that's GitHub's limitation, GitHub doesn't work that way. Since this PR is from a org (WinnerSoftLab), so no "Allow maintainer edit". Only the author can update the PR unless someone is added as the fork's collaborator.

I in fact sent a PR to that fork repository, they just need to review and merge it.

**Comment by silverwind** [comment]
Branch needs manual merge.

**Comment by KN4CK3R** [comment]
I think this PR just adds a "request for review" and not the ability to approve or merge (?)

**Comment by tacerus** [comment]
I see, thank you.

**Pull request merged**